### PR TITLE
added base profiler

### PR DESCRIPTION
**Pull request by kurisusnowdeng**

### CODE REVIEW

File: colossalai/utils/profiler/common.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from .comm_profiler import CommProfiler
+from typing import List
+from torch.utils.tensorboard import SummaryWriter
+
+
+class ProfilerEvent(Enum):
+    '''
+    Event types to profile
+    '''
+    COMM = CommProfiler
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:11]: why is profiler an event?

File: colossalai/utils/profiler/common.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+from enum import Enum
+from .comm_profiler import CommProfiler
+from typing import List
+from torch.utils.tensorboard import SummaryWriter
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:4]: should catch import error here as this requires tensorboard installation

File: colossalai/utils/profiler/_base_profiler.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+from torch.utils.tensorboard import SummaryWriter
+
+
+class BaseProfiler(object):
+
+    def __init__(self):
+        ...
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/_base_profiler.py:7]: this `...` will raise error

File: colossalai/utils/profiler/_base_profiler.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+from torch.utils.tensorboard import SummaryWriter
+
+
+class BaseProfiler(object):
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/_base_profiler.py:4]: make it an abstract class with abc

File: colossalai/utils/profiler/common.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        for e in self.events:
+            e.disable()
+
+    def to_tensorboard(self, writer: SummaryWriter):
+        for e in self.events:
+            e.to_tensorboard(writer)
+
+    def to_file(self, filename: str):
+        for e in self.events:
+            e.to_file(filename)
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:35]: multiple profilers write into the same file?

File: colossalai/utils/profiler/common.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    from torch.utils.tensorboard import SummaryWriter
+
+    with colossalai_profiler(events=[ProfilerEvent.COMM]) as prof:
+        train()
+    writer = SummaryWriter('tb/path')
+    prof.to_tensorboard(writer)
+    prof.to_file('./prof.log')
+    prof.show()
+    '''
+    return ProfilerContext(*args, **kwargs)
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:75]: This wrapper does not do anything, can just let user import ProfilerContext

Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:75]: Just mean to use lowercase api for context

File: colossalai/utils/profiler/common.py (hunk 5 above)
Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:35]: How about provide a path and add 'profilerevent_name.log' as a suffix

File: colossalai/utils/profiler/common.py (hunk 6 above)
Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:75]: It is alright to use `ProfilerContext`. In fact, the `*args` and `**kwargs` can hurt usability.

File: colossalai/utils/profiler/_base_profiler.py (hunk 3 above)
Comment (kurisusnowdeng) [reply colossalai/utils/profiler/_base_profiler.py:7]: Which error? I tried a simple test, and it was fine.

File: colossalai/utils/profiler/common.py (hunk 2 above)
Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:4]: Pytorch already helps raise the import error

File: colossalai/utils/profiler/_base_profiler.py (hunk 3 above)
Comment (FrankLeeeee) [reply colossalai/utils/profiler/_base_profiler.py:7]: Not an error, but this function should be removed if no initialization is specified.

File: colossalai/utils/profiler/_base_profiler.py (hunk 4 above)
Comment (feifeibear) [reply colossalai/utils/profiler/_base_profiler.py:4]: derive from ABC if it's an abstract class

File: colossalai/utils/profiler/common.py (hunk 1 above)
Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:11]: It is to distinguish events (or activities? as named by torch) to profile (e.g. communication, memory), and each event relates to a profiler

File: colossalai/utils/profiler/common.py (hunk 2 above)
Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:4]: This line makes TensorBoard a must.

Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:4]: Do you mean even there is no tensorboard installed we still make profiler work?

Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:4]: Yes

Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:4]: In fact, a better workaround of this is to import tensorboard only in `to_tensorboard`. Type hinting can be given as a string e.g. 

```python
writer: "torch.utils.tensorboard.SummaryWriter"
```

File: colossalai/utils/profiler/common.py (hunk 1 above)
Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:11]: It is kind of confusing as in communication profiler, it has defined its own event.

Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:11]: So change it to 'ProfilerActivity'?

File: colossalai/utils/profiler/common.py (hunk 5 above)
Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:35]: One way to handle this is to set the format as {output_path}/{prefix}_{profiler.__class__.__name__}_{timestamp}.log. This method requires changing the parameter filename to otuput_path

File: colossalai/utils/profiler/common.py (hunk 1 above)
Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:11]: Do not use abbreviation for API exposed to the user.

Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:11]: ok

Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:11]: It is not a common practice to put a class as the enum value.  It seems not standard to me even though it works. What do you think @feifeibear .

File: colossalai/utils/profiler/common.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+
+class ProfilerActivity(Enum):
+    '''
+    Event types to profile
+    '''
+    COMM = CommProfiler
+
+
+class colossalai_profiler(object):
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:17]: Follow python class naming convention

Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:17]: Follow python class naming convention

**Comment by kurisusnowdeng** [comment]
@FrankLeeeee All fixed.

File: colossalai/utils/profiler/__init__.py [hunk 8]
```diff
@@ -1 +1,2 @@
-from .comm_profiler import enable_communication_prof, communication_prof_show
+# from .comm_profiler import enable_communication_prof, communication_prof_show
+from .common import ProfilerActivity, colossalai_profiler
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/__init__.py:2]: Will this change cause the test_profiler unit testing to fail?

Comment (kurisusnowdeng) [reply colossalai/utils/profiler/__init__.py:2]: Definitely. Need to update unitests after the update of specific profilers.

File: colossalai/utils/profiler/common.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/utils/profiler/common.py:14]: Also weird to me

File: colossalai/utils/profiler/__init__.py (hunk 8 above)
Comment (kurisusnowdeng) [reply colossalai/utils/profiler/__init__.py:2]: Sure. Skip it.

File: colossalai/utils/profiler/common.py (hunk 7 above)
Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:17]: we can directly do this

```python
with Profiler(...) as prof:
    do_something()
```

I don't think we need to have the repository name in the class name.

Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:17]: Sure, but conventionally context classes are lowercase

Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py:17]: I see, I can accept the current code then.

**Comment by kurisusnowdeng** [comment]
@FrankLeeeee Fixed.

File: colossalai/utils/profiler/common.py (hunk 1 above)
Comment (FrankLeeeee) [reply colossalai/utils/profiler/common.py]: @kurisusnowdeng Any idea about this part?

Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py]: We can also use some keys to redirect to the profiler classes in a dict. But, I think this is a redundant step.

Comment (xcetos) [reply colossalai/utils/profiler/common.py]: Actually, using Enum as an option is quite common in C++ and Rust, but I have no idea whether this is appropriate for Python.

**Comment by FrankLeeeee** [comment]
Unit testing failed.

**Comment by kurisusnowdeng** [comment]
Updated. Should be fine now. Also changed the way to initialize the profiler context into passing profiler instances in, according to the suggestion from @1SAA and @xcetos 。

File: colossalai/utils/profiler/common.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+class ProfilerContext(object):
+    '''
+    Profiler context manager
+
+    Usage:
+    from colossalai.utils.profiler import ProfilerActivity, profiler
+    from torch.utils.tensorboard import SummaryWriter
+
+    with profiler(activities=[ProfilerActivity.COMMUNICATION]) as prof:
+        train()
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:18]: This should be updated.

File: colossalai/utils/profiler/common.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    writer = SummaryWriter('tb/path')
+    prof.to_tensorboard(writer)
+    prof.to_file('./prof_logs/')
+    prof.show()
+    '''
+
+    def __init__(self, activities: List[BaseProfiler] = None):
```
Comment (FrankLeeeee) [inline colossalai/utils/profiler/common.py:25]: shall we change the parameter name as the type is a profiler

Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:25]: I think 'activities' also makes sense, and maybe easier for pytorch users to distinguish the specific profilers from the profiler context (just too many 'profiler's).

Comment (kurisusnowdeng) [reply colossalai/utils/profiler/common.py:25]: @FrankLeeeee BTW, ready to merge if there is no other problems.

**Comment by FrankLeeeee** [comment]
According to @1SAA , this PR is replaced by #340 so I will close this PR now.

**Pull request closed without merging**